presto-main/src/main/java/com/facebook/presto/server/protocol/QueryBlockingRateLimiter.java

```diff
@@ -0,0 +1,96 @@
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.inject.Inject;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+/* Rate Limiting per query with token bucket
+ * Rate = rateLimitBucketMaxSize/second
+ *
+ * When having sufficient tokens, Request will be responded immediately.
+ * When not having enough tokens available, it uses the delayed processing method.
+ *
+ */
+public class QueryBlockingRateLimiter
+{
+    private final LoadingCache<QueryId, RateLimiter> rateLimiterCache;
+    private final long rateLimitBucketMaxSize;
+    private final ListeningExecutorService listeningExecutorService;
+
+    private CounterStat rateLimiterTriggeredCounter = new CounterStat();
+
+    @Inject
+    public QueryBlockingRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        checkArgument(rateLimitBucketMaxSize >= 1, "Rate limit token bucket max size should be at least 1");
+        // Using a custom thread pool with size 1-10 to reduce initial thread resources
+        ExecutorService executorService = new ThreadPoolExecutor(1, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), daemonThreadsNamed("rate-limiter-listener"));
+        listeningExecutorService = listeningDecorator(executorService);
+        rateLimiterCache = CacheBuilder.newBuilder().maximumSize(10000).expireAfterAccess(60, TimeUnit.MINUTES).build(CacheLoader.from(key -> RateLimiter.create(rateLimitBucketMaxSize)));
+    }
+
+    /*
+     * For accidental bug-caused DoS, we will use delayed processing method to reduce the requests, even when user do not have back-off logic implemented
+     * Optimized to avoid blocking for normal usages with TryRequire first
+     * Fall back to delayed processing method to acquire a permit, in a separate thread pool
+     * @Return:time spent sleeping to enforce rate, in seconds; 0.0 if not rate-limited
+     */
+    public ListenableFuture<Double> acquire(QueryId queryId)
+    {
+        requireNonNull(queryId, "queryId cannot be null");
```

Comment:
I would return a failed future here with IllegalArgumentException, so all error handling is contained in the future.

Reply:
will do